Fixes to `Sieve.primerange` and `primerange`

haru-44 opened the pull request:
* Added note to docstring.
  * `Sieve` cannot handle numbers greater than 2**31. This is a restriction due to the use of Type code `l` for `array`.
* Only odd numbers are sieved, excluding even numbers. This reduces execution time and memory usage by about half.
  * Further speedup can be achieved by excluding multiples of 3 and 5, but this was not implemented in this case.
* `primerange` previously used a list of known primes only if the given range was fully inclusive. Therefore;
  * If it is even partially covered, use `sieve`.
  * If it is less than or equal to the square of a known prime, the sieve is created (the prime obtained here is not recorded in `sieve`).


<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* ntheory
  * Added `sieve_interval` option to `Sieve`
<!-- END RELEASE NOTES -->

haru-44 committed:
Fixes to `Sieve.primerange` and `primerange`

* Added note to docstring.
  * `Sieve` cannot handle numbers greater than 2**31.
    This is a restriction due to the use of Type code `l` for `array`.
* Only odd numbers are sieved, excluding even numbers.
  This reduces execution time and memory usage by about half.
  * Further speedup can be achieved by excluding multiples of 3 and 5,
    but this was not implemented in this case.
* `primerange` previously used a list of known primes only
  if the given range was fully inclusive. Therefore;
  * If it is even partially covered, use `sieve`.
  * If it is less than or equal to the square of a known prime,
    the sieve is created (the prime obtained here is not recorded in `sieve`).

smichr commented on sympy/ntheory/generate.py:
```diff
@@ -39,7 +40,8 @@ class Sieve:
     growing sieve of Eratosthenes. When a lookup is requested involving
     an odd number that has not been sieved, the sieve is automatically
-    extended up to that number.
+    extended up to that number. Note that due to implementation problems,
+    it cannot handle more than 2**31-1.
```
```suggestion
    extended up to that number. Implementation details limit the number of
    primes to ``2^31-1``.
```

smichr commented on sympy/ntheory/generate.py:
```diff
@@ -745,14 +747,37 @@ def primerange(a, b=None):
+    if a <= largest_known_prime:
+        yield from sieve.primerange(a, largest_known_prime + 1)
+        a = largest_known_prime + 1
+    if a % 2:
```
Is this in case the largest_known_prime is 2?

haru-44 replied:
Implicitly assumes `2 < largest_known_prime`. This is because `_list[-1]` is never less than 13 in the `Sieve` constructor and `_reset`.

smichr replied:
Looking at it today, I think I was confusing `a` with something on the sieve; but `a` is just a given bound (which could be odd or even).

smichr commented on sympy/ntheory/generate.py:
```diff
@@ -745,14 +747,37 @@ def primerange(a, b=None):
+        a = largest_known_prime + 1
+    if a % 2:
+        a -= 1
+    # Array length to be allocated. The larger the length,
```
It seems like this must be duplicating something somewhere. Since this routine didn't formerly handle growing the sieve, why does it need to do so now? Should this growth mechanism be abstracted to a method for the sieve itself rather than be part of the `primerange` function?

smichr replied:
That makes sense to me. I am in favor of that change.

smichr commented:
Just a few questions and suggestions added here.

haru-44 committed:
Update sympy/ntheory/generate.py

Co-authored-by: Christopher Smith <smichr@gmail.com>

haru-44 committed:
Update sympy/ntheory/generate.py

Co-authored-by: Christopher Smith <smichr@gmail.com>

haru-44 commented:
Thanks for the review. If there are no problems, I would like to make the above modifications.

haru-44 committed:
Added `sieve_interval` option to `Sieve`

* The type code of `Sieve._list` has been changed to `L`.
  This allows prime numbers up to `2**32-1` to be handled.
* A new `Sieve._primerange` is created. Here we do the
  actual sieve extending, and `Sieve.extend` only merges
  the obtained results into the `_list`.
* `primerange` can be used with `Sieve._primerange` to efficiently
  obtain prime numbers.

haru-44 commented:
I created the new function. The assumptions are quite unique and should be used with caution. I thought I should write more notes, but what should I do?

haru-44 committed:
fix doctest

smichr commented on sympy/ntheory/tests/test_generate.py:
```diff
@@ -142,6 +148,21 @@ def test_generate():
 
+    for sieve_interval in [1, 10, 11, 1_000_000]:
+        s = Sieve(sieve_interval=sieve_interval)
+        assert s._list[-1] == 13
```
Do I understand correctly that `sieve_interval` is the size by which we let the sieve grow (but it always starts with 6 elements, the last being 13)?

haru-44 replied:
That's right. The `sieve_interval` is the length of the array to allocate in one sieve. Normally, it should be set to some large value, but setting it to `1` works correctly in theory.

The `s._list[-1] == 13` check was intended to make the precondition explicit. If the default value of `_list` were changed to `[2, 3, 5, 7, 11]`, the program would still work correctly, but would no longer satisfy the test condition. This is misleading and has been fixed.

smichr commented on sympy/ntheory/generate.py:
```diff
@@ -55,11 +55,28 @@ class Sieve:
         self._n = 6
-        self._list = _aset(2, 3, 5, 7, 11, 13) # primes
+        self._list = _array('L', [2, 3, 5, 7, 11, 13]) # primes
         self._tlist = _aset(0, 1, 1, 2, 2, 4) # totient
         self._mlist = _aset(0, 1, -1, -1, 0, -1) # mobius
```
tlist could be array type "L" and mlist array type "i". Then `_aset` could be removed and `_arange` changed to use type "L"

haru-44 replied:
Remove `_azeros`, `_aset`, and `_arange`, since it is no longer worthwhile to have them as functions.

smichr commented on sympy/ntheory/generate.py:
```diff
@@ -745,14 +747,37 @@ def primerange(a, b=None):
         return
+    # If we know some of it, return it.
+    if a <= largest_known_prime:
+        yield from sieve.primerange(a, largest_known_prime + 1)
```
`bisect` to find starting point, you mean? That could be internal to `primerange`.

smichr replied:
I think this is ok as is. It is just a highlevel call to sieve's primerange which, in turn, will use bisection to find the right starting position.

haru-44 committed:
removed `_azeros`, `_aset`, `_arange`

haru-44 commented:
Is there anything I can do?

haru-44 commented:
Thank you. By the way, I know the python 3.12 test fails, but why is the relace-note not running?

smichr commented:
restarted tests

smichr commented:
I think it is the sphynx test that is holding this up. Maybe try rebase with current master?

haru-44 commented:
Was this the correct way to rebase? (I have never done rebase before).

smichr commented:
I don't know which test is holding this up. And the rebase looks fine.

haru-44 commented:
The sphinx error seems to have been resolved in #25249. Do I need to rebase again?

haru-44 commented:
@smichr 
If everything is in order, could you kindly proceed with the merge, please?

haru-44 commented:
Thank you.
